mlflow/sklearn/__init__.py

```diff
@@ -884,12 +941,69 @@ def patched_fit(original, self, *args, **kwargs):
                           for autologging (e.g., specify "fit" in order to indicate that
                           `sklearn.linear_model.LogisticRegression.fit()` is being patched)
         """
+        global _autolog_training_status
+        with _SklearnTrainingSession(clazz=self.__class__, allow_children=False) as t:
+            if t.should_log():
+                _autolog_training_status = _AutologTrainingStatus(
+                    last_mlflow_run_id=mlflow.active_run().info.run_id,
+                    model_id=id(self)
+                )
+                print(f'DBG: run patched_fit,'
+                      f'run_id={_autolog_training_status.last_mlflow_run_id},'
+                      f'model_id={_autolog_training_status.model_id}.')
+                try:
+                    _autolog_training_status.in_fit_call_scope = True
+                    result = fit_mlflow(original, self, *args, **kwargs)
+                finally:
+                    _autolog_training_status.in_fit_call_scope = False
+                return result
+            else:
+                return original(self, *args, **kwargs)
+
+    def patched_predict(original, self, *args, **kwargs):
+        from mlflow.utils import _inspect_original_var_name
+        global _autolog_training_status
         with _SklearnTrainingSession(clazz=self.__class__, allow_children=False) as t:
             if t.should_log():
-                return fit_mlflow(original, self, *args, **kwargs)
+                result = original(self, *args, **kwargs)
+                eval_dataset = args[0] if len(args) >= 1 else kwargs.get('X')
+                eval_dataset_name = _inspect_original_var_name(eval_dataset)
+                _autolog_training_status.register_predition_result(
+                    self, eval_dataset, result, eval_dataset_name)
+                return result
             else:
                 return original(self, *args, **kwargs)
 
+    def log_eval_metric(metric_name, metric, metric_api_call_arg_list):
+        with ResumeAutologRun() as run:
+            print(f'DBG: run patched_metric_api #3, resume run_id={run.info.run_id}')
+            eval_dataset_name = _autolog_training_status.get_eval_dataset_name_from_metric_api_arglist(
+                metric_api_call_arg_list
+            )
+            if eval_dataset_name:
+                print('DGB: run patched_metric_api #4')
+                # TODO: handle the case log the same metric key multiple times.
+                mlflow.log_metric(
+                    f'{metric_name}_on_{eval_dataset_name}',
+                    metric
+                )
+
+    def patched_metric_api(original, *args, **kwargs):
+        # TODO:
+        #  check _SklearnTrainingSession and skip logging if in nested runs.
+        global _autolog_training_status
+        metric = original(*args, **kwargs)
```

Comment:
Do all metrics APIs return a scalar value?

Reply:
We can consider skip logging such metric api

Reply:
Or we could log `confusion_matrix` as an arifact (and do the same thing for other metrics)

Reply:
I'd vote for logging only scalar metrics in the first version.